### PR TITLE
Batch-based processing filter_data.py that calculates mean gradients

### PR DESCRIPTION
**Pull request by dipikakhullar**
1. Batch-based processing system that calculates mean gradients, normalizes gradients, and computes inner products in smaller batches…zes gradients, and computes inner products in smaller batches

### CODE REVIEW

File: examples/filter_data.py [hunk 1]
```diff
@@ -125,15 +122,33 @@ def main(
...
+                    mean_gradient = batch_gradients.mean(0)
+                else:
+                    mean_gradient = (mean_gradient * i + batch_gradients.sum(0)) / (i + len(batch))
+                
+                # Clear memory
+                del batch_gradients
+            
+            query = mean_gradient.cuda()
+            del query_index, mean_gradient
+            torch.cuda.empty_cache()
```
Comment (luciaquirke) [inline examples/filter_data.py:150]: Many people in the alignment community add the `torch.cuda.empty_cache()` line when they're having memory leaks they don't know how to fix, but it doesn't fix the leaks or achieve anything in most contexts. They are not setting a good example, plz do not adopt 😅 

My understanding is that `torch.cuda.empty_cache()` releases VRAM reserved by PyTorch that it doesn't need anymore, so if you are simultaneously running other processes that need VRAM it can be helpful.

File: examples/filter_data.py [hunk 2]
```diff
@@ -191,30 +246,42 @@ def add_rank(ex):
...
             raise ValueError(f"Invalid filter: {args.filter}")
 
+        # Clear memory before loading model
+        torch.cuda.empty_cache()
+        
+        print("Loading model...")
         model = AutoModelForCausalLM.from_pretrained(
             args.model,
             torch_dtype="bfloat16",
+            low_cpu_mem_usage=True,
```
Comment (luciaquirke) [inline examples/filter_data.py:264]: Does this do anything? Seems deprecated

File: examples/filter_data.py [hunk 3]
```diff
@@ -81,7 +82,7 @@ def select_topk(ds: Dataset, n: int, key: str, lowest: bool = False):
 
 def normalize_batch(batch):
```
Comment (luciaquirke) [inline examples/filter_data.py:83]: Can be removed

File: examples/filter_data.py [hunk 4]
```diff
@@ -125,15 +122,33 @@ def main(
...
+                batch = query_index.select(range(i, min(i + batch_size, len(query_index))))
+                batch_gradients = torch.stack([g for g in batch["gradients"]])
+                
+                # Normalize gradients
+                batch_gradients = F.normalize(batch_gradients, dim=1)
+                
+                if mean_gradient is None:
+                    mean_gradient = batch_gradients.mean(0)
+                else:
+                    mean_gradient = (mean_gradient * i + batch_gradients.sum(0)) / (i + len(batch))
```
Comment (luciaquirke) [inline examples/filter_data.py:143]: Mean calculation is incorrect

File: examples/filter_data.py [hunk 5]
```diff
@@ -125,15 +122,33 @@ def main(
...
+            print("Computing query mean gradient...")
+            mean_gradient = None
+            batch_size = 512
+            
+            for i in tqdm(range(0, len(query_index), batch_size), desc="Computing query mean"):
+                batch = query_index.select(range(i, min(i + batch_size, len(query_index))))
+                batch_gradients = torch.stack([g for g in batch["gradients"]])
+                
+                # Normalize gradients
+                batch_gradients = F.normalize(batch_gradients, dim=1)
```
Comment (luciaquirke) [inline examples/filter_data.py:138]: To get performance benefits from this method we need to move each batch to CUDA

File: examples/filter_data.py [hunk 6]
```diff
@@ -160,20 +175,60 @@ def main(
             pass
         elif args.filter == "attribution":
             if query == None:
-                query = train["gradient"].mean(0).cuda()
+                # Calculate mean gradient in batches to avoid memory issues
+                print("Calculating mean gradient...")
+                mean_gradient = None
+                batch_size = 100  # Process in smaller batches
```
Comment (luciaquirke) [inline examples/filter_data.py:181]: 512, I reckon also add the batch size to FilterConfig like we talked about

File: examples/filter_data.py [hunk 7]
```diff
@@ -160,20 +175,60 @@ def main(
...
-                Dataset, Dataset.from_generator(importance_score_generator)
-            )
-            train = select_topk(train, args.num_examples, "importance_score", lowest=args.lowest)
+            print("Computing importance scores...")
+            # Calculate importance scores in batches
+            importance_scores = []
+            batch_size = 512
+            
+            for i in tqdm(range(0, len(train), batch_size), desc="Computing importance scores"):
+                batch = train.select(range(i, min(i + batch_size, len(train))))
```
Comment (luciaquirke) [inline examples/filter_data.py:203]: I _think_ you can do `train[i : min(i + batch_size, len(train))]` then remove the `torch.stack`

File: examples/filter_data.py [hunk 8]
```diff
@@ -160,20 +175,60 @@ def main(
...
-                        "importance_score": (row["gradient"].cuda() @ query).item(),
-                    }
-
-            train = assert_type(
-                Dataset, Dataset.from_generator(importance_score_generator)
-            )
-            train = select_topk(train, args.num_examples, "importance_score", lowest=args.lowest)
+            print("Computing importance scores...")
+            # Calculate importance scores in batches
+            importance_scores = []
```
Comment (luciaquirke) [inline examples/filter_data.py:199]: Preallocate torch.zeros vector of len(train)

File: examples/filter_data.py [hunk 9]
```diff
@@ -160,20 +175,60 @@ def main(
...
+            for i in tqdm(range(0, len(train), batch_size), desc="Computing importance scores"):
+                batch = train.select(range(i, min(i + batch_size, len(train))))
+                batch_gradients = torch.stack([g for g in batch["gradients"]]).cuda()
+                
+                # Normalize gradients
+                batch_gradients = F.normalize(batch_gradients, dim=1)
+                
+                # Compute inner products
+                batch_scores = (batch_gradients @ query).cpu().numpy()
+                importance_scores.extend(batch_scores)
```
Comment (luciaquirke) [inline examples/filter_data.py:211]: Add to tensor using the same indexing as for the batch selection

File: examples/filter_data.py [hunk 10]
```diff
@@ -191,30 +246,42 @@ def add_rank(ex):
...
+        print("Loading model...")
         model = AutoModelForCausalLM.from_pretrained(
             args.model,
             torch_dtype="bfloat16",
+            low_cpu_mem_usage=True,
         )
         tokenizer = AutoTokenizer.from_pretrained(args.model, max_length=8192)
         model, tokenizer = setup_chat_format(model, tokenizer)
 
+        # Create DataConfig for tokenization
```
Comment (luciaquirke) [inline examples/filter_data.py:269]: Nice

File: examples/filter_data.py (hunk 4 above)
Comment (luciaquirke) [reply examples/filter_data.py:143]: If you don't have cursor autocomplete enabled I recommend enabling it cause it should be able to help with this
